Add Steam game ID and screenshot paths as attributes

i00 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
None

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Changes have been added to add extra attributes for the Steam game ID & game image URLs to the Steam sensor.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
# No changes from base integration
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
### Example
This example will display the currently playing game image in lovelace:
```yaml
type: markdown
content: >-
  <img src = '{{
  states.sensor.STEAM_ENTITY_ID.attributes.game_image_main }}'
  width=100%>
```

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: **No documentation changes required**

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

**No documentation changes required**
- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

**Not required**
- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

i00 committed:
Merge from Upstream

i00 committed:
Update Steam intergration to list screenshot paths and Steam game ID

i00 commented:
Can someone explain to me why these checks are failing ...?
The changes look the same as the rest of the code to me ... also it works in HA!

Thanks

dshokouhi commented:
@i00 did you run the pre-commit setup when you created the PR?  I assume you followed the steps here to get your dev environment setup?  https://developers.home-assistant.io/docs/development_environment


This suggests what you may need to run to recommit:
https://dev.azure.com/home-assistant/Core/_build/results?buildId=28096&view=logs&j=2fa2c639-d05e-530d-034e-4fb413fcea4e&t=88bd70b9-15d8-56b4-acb9-94f6fcce0e9d&l=35

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -170,6 +172,10 @@ def device_state_attributes(self):
+        if self._game_id is not None:
+            attr["game_id"] = self._game_id
+            attr["game_image_header"] = f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/header.jpg"
+            attr["game_image_main"] = f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/capsule_616x353.jpg"
```
Wouldn't it be useful to utilize [entity_picture](https://developers.home-assistant.io/docs/entity_index/#generic-properties) for one of this pictures?

Please create a constant for `https://steamcdn-a.akamaihd.net/steam/apps/` to reuse it.

i00 replied:
Good point .. will do thanks :)

i00 replied:
Done both of those changes you suggested; I used the header as the entity_picture (as it is quite small) when in-game, and the profile picture when not.

i00 commented:
@dshokouhi

> @i00 did you run the pre-commit setup when you created the PR? I assume you followed the steps here to get your dev environment setup? https://developers.home-assistant.io/docs/development_environment

Yes I did follow those instructions

> This suggests what you may need to run to recommit:
> https://dev.azure.com/home-assistant/Core/_build/results?buildId=28096&view=logs&j=2fa2c639-d05e-530d-034e-4fb413fcea4e&t=88bd70b9-15d8-56b4-acb9-94f6fcce0e9d&l=35

... the error doesn't mean much to me ... how do I see why black failed?

Thanks

dshokouhi commented:
> how do I see why black failed?

@i00 Its actually on the same page I linked.  At the very bottom line 57 it says check black formatting.  Up above you can see the changes that black wants to make, line 45-52.  The line that I linked to does a pre-commit check and makes the modifications from line 45-52 before you commit which would avoid this error.  When you setup your development environment it should be installed when you run `script/setup` https://developers.home-assistant.io/docs/development_environment#setting-up-virtual-environment

i00 commented:
@dshokouhi 
Ok ... I have the thing in a Linux docker container... and when I commit through my Windows box it ignores the hooks  ...

This is the output when I run a git commit through the Linux command line:

```
black....................................................................Failed
- hook id: black
- files were modified by this hook
codespell................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

homeassistant/components/steam_online/sensor.py:170:1: D400 First line should end with a period

bandit...................................................................Passed
isort....................................................................Passed
Check JSON...........................................(no files to check)Skipped
mypy.....................................................................Passed
root@debian1:/home-assistant# git status
On branch kb_Steam
Your branch is up to date with 'origin/kb_Steam'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   homeassistant/components/steam_online/sensor.py

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   homeassistant/components/steam_online/sensor.py
```

...The thing that strikes me as strange is that it changes:
```python
"""This:"""
            return f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/header.jpg"
"""To:"""
            return (
                f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/header.jpg"
            )

"""This:"""
            attr["game_image_main"] = f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/capsule_616x353.jpg"
"""To:"""
            attr[
                "game_image_main"
            ] = f"https://steamcdn-a.akamaihd.net/steam/apps/{self._game_id}/capsule_616x353.jpg"
```

How on earth is that an improvement??!

i00 commented:
@dshokouhi ...
So I let code black make the file more illegible :P... and now flake is failing ... with exit code: 1 ... how do I check that?

Thanks

dshokouhi commented:
Black is the adopted code formatter https://developers.home-assistant.io/blog/2019/07/31/black/

Regarding Flake8, it looks like it tells you whats wrong just beneath that error.  So you would need to resolve that as well so you get the checks to pass.  I havent experienced this error before so you may need to research it if its not as simple as it says.

i00 committed:
Steam entity_picture now changed to game art when in-game

i00 commented:
@dshokouhi .. thanks an easy fix (comment didn't end in a fullstop).. don't know how I missed that in the log ... checks happening now  *cross fingers* 👍

i00 committed:
Steam - changing API endpoints to constants

i00 committed:
Steam - formatting code for lint

springstan reviewed (commented):
I did not know that the steam integration was already using this property, do you think it is meaningful to switch between the current game image and the avatar image?
Considering that your previous solution was already accessible and usable in Lovelace.

i00 commented:
@springstan - I actually quite like it doing this ... maybe I should make it an option in settings?
If so should it switch by default to the game art when playing - opt in ... or not opt out?

Also what should I call the setting? I am thinking something like "switch_entity_picture_when_in_game" ... kind of long :P ... but fairly obvious as to what it does.

springstan commented:
I understand your personal use case, but it does not align with other integrations and I would personally prefer to preserve that consistency. Furthermore, now the image of the personal avatar is unreachable while playing a game which is unusual as well.
Consequently, IMHO it should be implemented as two attributes instead of switching the entity picture based on the current activity. Therefore, I would not suggest to create an opt in or opt out config option.

i00 committed:
Update sensor.py

Removing entity_picture image switching based on current playing game

i00 commented:
@springstan ... ok I have removed the entity_picture image switching based on current playing game :)

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -28,6 +28,9 @@
 STATE_LOOKING_TO_PLAY = "looking_to_play"
 
+STEAM_API_URL = "https://steamcdn-a.akamaihd.net/steam/"
+STEAM_APP_ENDPOINT = "apps/"
```
Combine these two constants into one:
```suggestion
STEAM_API_URL = "https://steamcdn-a.akamaihd.net/steam/apps/"
```

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -164,12 +169,25 @@ def _get_last_online(self):
+        if self._game_id is not None:
+            return f"{STEAM_API_URL}{STEAM_APP_ENDPOINT}{self._game_id}/header.jpg"
+
+        return None
```
Remove this entire method and move its functionality to `device_state_attributes()`.

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -164,12 +169,25 @@ def _get_last_online(self):
+    def _get_game_image_header(self):
+        """Return the current games header image URL."""
+        if self._game_id is not None:
+            return f"{STEAM_API_URL}{STEAM_APP_ENDPOINT}{self._game_id}/header.jpg"
```
Please create constants for `header.jpg` and `capsule_616x353.jpg`.

i00 commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -104,6 +108,7 @@ def update(self):
         try:
             self._profile = self._steamod.user.profile(self._account)
             self._game = self._get_current_game()
+            self._game_id = self._profile.current_game[0]
```
Well spotted :)

i00 committed:
Steam - tidying up code

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -170,6 +177,14 @@ def device_state_attributes(self):
+            ] = f"{STEAM_API_URL}{self._game_id}/{STEAM_HEADER_IMAGE_FILE}"
+            attr[
+                "game_image_main"
+            ] = f"{STEAM_API_URL}{self._game_id}/{STEAM_MAIN_IMAGE_FILE}"
```
Introduce a local variable for the url including the game id:
```suggestion
            game_url = f"{STEAM_API_URL}{self._game_id}/"
            attr["game_image_header"] = f"{game_url}{STEAM_HEADER_IMAGE_FILE}"
            attr["game_image_main"] = f"{game_url}{STEAM_MAIN_IMAGE_FILE}"
```

springstan commented on homeassistant/components/steam_online/sensor.py:
```diff
@@ -104,6 +109,7 @@ def update(self):
         try:
             self._profile = self._steamod.user.profile(self._account)
             self._game = self._get_current_game()
+            self._game_id = self._profile.current_game[0]
```
Does `self._profile.current_game[0]` return `None` when you are not currently playing a game?

i00 replied:
I am not 100% sure to be honest ...
I never bothered checking it since the call to self._get_current_game() (line above) gets self._profile.current_game[0] ... so I assumed that this isn't a dodge try catch to handle that.. will check that out tomorrow.

springstan replied:
Any updates on this? :)

i00 replied:
Sorry ... had other commitments crop up :( ... checking now :)

i00 replied:
Yes it does .. I thought that the "self._profile.current_game" may have been failing since it may be None when not playing (would have been swallowed by the exception if this was happening)... but instead it contains 3 None's :) ... so looks all good 👍

springstan replied:
Perfect thanks for looking into it :)

springstan reviewed (commented):
Just two small things that need to be addressed. 
Btw could you please update the documentation for this integration? Thanks 👍

i00 commented:
@springstan ..
Sorry to ask but why does the documentation require updating? ...
The config has not changed, there are only extra attributes added to the entity.

It is rare that any details about what is actually is returned in a sensor appear in the documentation; the current [Steam documentation](https://www.home-assistant.io/integrations/steam_online/) does not list last_online, and level for instance.

springstan commented:
@i00 yeah I did not mean to add these attributes, I thought you wanted to add your example with the lovelace markdown card to the docs to show whtat is possible :)

i00 commented:
I have a panel page setup with just the markdown image on it.
When my Android TV is on Steam Link I cast the page to the Google Nest Hub next to my TV.
As I turn off the TV or exit Steam Link it stops casting.

But it is really there so people can use it how they want :). I'll leave it up to you @springstan.

springstan commented:
You convinced me @i00, it is fine how it currently is :)

i00 commented:
So how do we go about removing the docs-missing check?
Thanks

springstan commented:
The docs-missing check is a bit delayed to removing the label but is automatically removed after some time.

i00 commented:
Any update on this @springstan .. still has the docs missing.. has been four days now?.. should I revert and re-commit?

springstan commented:
I think it is fine and hopefully will disappear with another commit. Could you please take a look at the changes I requested? Thanks :)

i00 committed:
Update homeassistant/components/steam_online/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

springstan reviewed (approved):
LGTM 👍
